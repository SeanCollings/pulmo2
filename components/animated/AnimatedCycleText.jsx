import React, { useEffect, useState } from 'react';
import { StyleSheet, View, Text, Animated } from 'react-native';
import { useTheme } from '../../hooks/useTheme';

const AnimatedCycleText = ({
  textArray = [],
  fadeIn = 2000,
  fadeOut = 1000,
  delay = 0,
  repeat = false,
}) => {
  const theme = useTheme();
  const [animatedOpacities] = useState(
    textArray.map(() => new Animated.Value(0))
  );

  if (!textArray.length) return null;

  const maxOpacity = theme.DARK ? 0.87 : 1;
  const getAnimations = () =>
    textArray.reduce((acc, _, i) => {
      const keepLastVisible = !repeat && i === textArray.length - 1;

      if (keepLastVisible) {
        return [
          ...acc,
          Animated.timing(animatedOpacities[i], {
            toValue: maxOpacity,
            duration: fadeIn,
          }),
        ];
      } else {
        return [
          ...acc,
          Animated.timing(animatedOpacities[i], {
            toValue: maxOpacity,
            duration: fadeIn,
          }),
          Animated.delay(delay),
          Animated.timing(animatedOpacities[i], {
            toValue: 0,
            duration: fadeOut,
          }),
        ];
      }
    }, []);

  const opacityAnimations = getAnimations();

  useEffect(() => {
    if (repeat) {
      Animated.loop(Animated.sequence(opacityAnimations)).start();
    } else {
      Animated.sequence(opacityAnimations).start();
    }
  }, []);

  return (
    <View style={styles.container}>
      {textArray.map((text, i) => {
        return (
          <Animated.Text
            key={`animted-${i}`}
            style={{
              ...styles.textStyle,
              color: theme.TEXT,
              opacity: animatedOpacities[i],
            }}
          >
            {text}
          </Animated.Text>
        );
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  container: { alignItems: 'center', height: 60 },
  textStyle: {
    fontSize: 40,
    textAlignVertical: 'center',
    fontFamily: 'tit-light',
    textTransform: 'lowercase',
    position: 'absolute',
  },
});

export default AnimatedCycleText;
