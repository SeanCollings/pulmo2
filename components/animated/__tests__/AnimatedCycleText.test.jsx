import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup } from 'react-native-testing-library';
import AnimatedCycleText from '../AnimatedCycleText';

jest.useFakeTimers();

describe('AnimatedAppear - unit test', () => {
  let component;
  afterEach(() => {
    cleanup();
  });

  test('should return null if empty text array', async () => {
    await act(async () => {
      component = create(<AnimatedCycleText />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should render default with a non-empty text array of length 1', async () => {
    await act(async () => {
      component = create(
        <AnimatedCycleText
          textArray={['AnimatedCycleText']}
          fadeIn={0}
          fadeOut={0}
        />
      );
    });
    expect(component).toMatchSnapshot();
  });

  test('should render with multiple text array values', async () => {
    await act(async () => {
      component = create(
        <AnimatedCycleText
          textArray={['one', 'two', 'three']}
          fadeIn={0}
          fadeOut={0}
        />
      );
    });
    expect(component).toMatchSnapshot();
  });
});
