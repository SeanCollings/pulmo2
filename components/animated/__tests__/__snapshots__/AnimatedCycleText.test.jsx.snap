// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedAppear - unit test should render default with a non-empty text array of length 1 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 60,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#3e4b57",
        "fontFamily": "tit-light",
        "fontSize": 40,
        "opacity": 1,
        "position": "absolute",
        "textAlignVertical": "center",
        "textTransform": "lowercase",
      }
    }
  >
    AnimatedCycleText
  </Text>
</View>
`;

exports[`AnimatedAppear - unit test should render with multiple text array values 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 60,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#3e4b57",
        "fontFamily": "tit-light",
        "fontSize": 40,
        "opacity": 0,
        "position": "absolute",
        "textAlignVertical": "center",
        "textTransform": "lowercase",
      }
    }
  >
    one
  </Text>
  <Text
    style={
      Object {
        "color": "#3e4b57",
        "fontFamily": "tit-light",
        "fontSize": 40,
        "opacity": 0,
        "position": "absolute",
        "textAlignVertical": "center",
        "textTransform": "lowercase",
      }
    }
  >
    two
  </Text>
  <Text
    style={
      Object {
        "color": "#3e4b57",
        "fontFamily": "tit-light",
        "fontSize": 40,
        "opacity": 1,
        "position": "absolute",
        "textAlignVertical": "center",
        "textTransform": "lowercase",
      }
    }
  >
    three
  </Text>
</View>
`;

exports[`AnimatedAppear - unit test should return null if empty text array 1`] = `null`;
